wrc/codegen/merger.py

```diff
@@ -0,0 +1,179 @@
+    if not isinstance(astreg.sections[NOTES_INDEX], Section) or \
+            not isinstance(astguide.sections[NOTES_INDEX], Section):
+        raise BadFormatError('NOTES_INDEX is incorrect')
+    if not isinstance(astreg.sections[TOC_INDEX], TableOfContent):
+        raise BadFormatError('TOC_INDEX is incorrect')
+    for document in [astreg, astguide]:
+        if not hasattr(document.sections[FIRST_ARTICLE_INDEX], 'number') or \
+                document.sections[FIRST_ARTICLE_INDEX].number != '1':
+            raise BadFormatError('FIRST_ARTICLE_INDEX is incorrect')
+
+    # We take the astreg as the base, and then we add the guidelines in their corresponding places.
+    ast_combined = astreg
+
+    # Merge the notes:
+    # We won't include the "WCA Regulations" notes from the Guidelines, as it is the same as in
+    # "WCA Regulations and Guidelines" of the Regulations document.
+    # languages_options is used to handle translations.
+    ast_combined.sections[NOTES_INDEX].content.append(
+        [cont for cont in astguide.sections[NOTES_INDEX].content if language_options['regulations'] not in cont.title])
+
+    # Now we iterate over all the Articles found in astguide, from FIRST_ARTICLE_INDEX until the end of the
+    # sections list.
+    guide_sections = astguide.sections[FIRST_ARTICLE_INDEX:]
+    for guide_section in guide_sections:
+
+        stack = deepcopy(guide_section.content)
+        reg_section = None
+        # Find the associated section in the Regulations or add it if there isn't one.
+        ast_combined_sections = ast_combined.sections[FIRST_ARTICLE_INDEX:]
+        for index, section in enumerate(ast_combined_sections):
+            if not isinstance(section, Article):
+                # We could just continue here, but finding a non-Article section after FIRST_ARTICLE_INDEX
+                # is worth raising an exception.
+                raise BadFormatError('Section is not an Article')
+
+            if section.number == guide_section.number:
+                # We found the section!
+                reg_section = section
+                break
+            elif article_number_is_lower(guide_section.number, section.number):
+                # This happens when an Article in the Guidelines doesn't have an associated article in the Regulations.
+                ast_combined.sections[TOC_INDEX].articles.insert(index, guide_section)
+                # Now we copy the Article from the Guidelines into the ast_combined and continue (empty stack).
+                ast_combined.sections.insert(index + FIRST_ARTICLE_INDEX, guide_section)
+                stack = []
+                break
+
+        # Insert guidelines until the stack is empty.
+        while len(stack) > 0:
+            guideline = stack.pop(0)
+            guideline_num = get_reg_number(guideline)
+            inserted_index = recursive_insert(reg_section.content, guideline, guideline_num, stack)
+            assert inserted_index != -1
```

Comment:
Does this raise a hard error (i.e. terminate program execution) when the assertion is `False` or does it "only" print a warning message and try to continue?

Reply:
It raises a hard error. Anyway, I changed it to:
```python3
if inserted_index == -1:
    raise BadFormatError(f'Could not insert guideline {guideline.number}')
```